src/ledger/LedgerState.cpp

```diff
@@ -0,0 +1,1311 @@
+            (oe.selling == asset || oe.buying == asset))
+        {
+            offers[key] = *entry;
+        }
+    }
+    return offers;
+}
+
+LedgerStateEntry
+LedgerState::load(LedgerKey const& key)
+{
+    return mImpl->load(*this, key);
+}
+
+LedgerStateEntry
+LedgerState::Impl::load(LedgerState& self, LedgerKey const& key)
+{
+    checkNotSealed();
+    checkNoChild();
+    if (mActive.find(key) != mActive.end())
+    {
+        throw std::runtime_error("Key already loaded");
+    }
+
+    auto newest = getNewestVersion(self, key);
+    if (!newest.entry)
+    {
+        return {};
+    }
+
+    auto current = std::make_shared<LedgerEntry>(*newest.entry);
+    mEntry[key] = current;
+
+    auto impl = std::make_shared<LedgerStateEntryImpl>(self, *current);
+    mActive.emplace(key, impl);
+    return LedgerStateEntry(impl);
+}
+
+std::map<AccountID, std::vector<LedgerStateEntry>>
+LedgerState::loadAllOffers()
+{
+    return mImpl->loadAllOffers(*this);
+}
+
+std::map<AccountID, std::vector<LedgerStateEntry>>
+LedgerState::Impl::loadAllOffers(LedgerState& self)
+{
+    auto offers = getAllOffers();
+    std::map<AccountID, std::vector<LedgerStateEntry>> offersByAccount;
+    for (auto const& kv : offers)
+    {
+        auto const& key = kv.first;
+        auto const& sellerID = key.offer().sellerID;
+        offersByAccount[sellerID].emplace_back(load(self, key));
+    }
+    return offersByAccount;
+}
+
+LedgerStateEntry
+LedgerState::loadBestOffer(Asset const& buying, Asset const& selling)
+{
+    return mImpl->loadBestOffer(*this, buying, selling);
+}
+
+LedgerStateEntry
+LedgerState::Impl::loadBestOffer(LedgerState& self, Asset const& buying,
+                                 Asset const& selling)
+{
+    checkNotSealed();
+    checkNoChild();
+
+    auto le = getBestOffer(buying, selling, {});
+    return le ? load(self, LedgerEntryKey(*le)) : LedgerStateEntry();
+}
+
+LedgerStateHeader
+LedgerState::loadHeader()
+{
+    return mImpl->loadHeader(*this);
+}
+
+LedgerStateHeader
+LedgerState::Impl::loadHeader(LedgerState& self)
+{
+    checkNotSealed();
+    checkNoChild();
+    if (mActiveHeader)
+    {
+        throw std::runtime_error("LedgerStateHeader is active");
+    }
+
+    mActiveHeader = std::make_shared<HeaderImpl>(self, mHeader);
+    return LedgerStateHeader(mActiveHeader);
+}
+
+std::vector<LedgerStateEntry>
+LedgerState::loadOffersByAccountAndAsset(AccountID const& accountID,
+                                         Asset const& asset)
+{
+    return mImpl->loadOffersByAccountAndAsset(*this, accountID, asset);
+}
+
+std::vector<LedgerStateEntry>
+LedgerState::Impl::loadOffersByAccountAndAsset(LedgerState& self,
+                                               AccountID const& accountID,
+                                               Asset const& asset)
+{
+    checkNotSealed();
+    checkNoChild();
+
+    std::vector<LedgerStateEntry> res;
+    auto offers = getOffersByAccountAndAsset(accountID, asset);
```

Comment:
It looks like:

LedgerState::loadOffersByAccountAndAsset calls
LedgerState::Impl::loadOffersByAccountAndAsset which calls
LedgerState::Impl::getOffersByAccountAndAsset

but:

LedgerStateRoot::getOffersByAccountAndAsset calls
LedgerStateRoot::Impl::getOffersByAccountAndAsset which callsLedgerStateRoot::Impl::loadOffersByAccountAndAsset

I think that this load/get relationship is not 100% clear for me.
Could you explain that?